src/python/pants/core/util_rules/gen_rule_test.py

```diff
@@ -0,0 +1,154 @@
+) -> None:
+    target = rule_runner.get_target(address)
+    result = rule_runner.request(
+        GeneratedSources, [GenerateFilesFromGenRuleRequest(EMPTY_SNAPSHOT, target)]
+    )
+    assert result.snapshot.files == tuple(expected_contents)
+    contents = rule_runner.request(DigestContents, [result.snapshot.digest])
+    for fc in contents:
+        assert fc.content == expected_contents[fc.path].encode()
+
+
+def test_sources_and_files(rule_runner: RuleRunner) -> None:
+    MSG = ["Hello gen_rule", ", nice cut."]
+    rule_runner.write_files(
+        {
+            "src/BUILD": dedent(
+                """\
+                gen_rule(
+                  name="hello",
+                  sources=["script"],
+                  dependencies=[":files"],
+                  tools=[
+                    "cat",
+                    "mkdir",
+                    "tee",
+                  ],
+                  outputs=["message.txt", "res/"],
+                  command="source ./script",
+                )
+
+                files(
+                  name="files",
+                  sources=["*.txt"],
+                )
+                """
+            ),
+            "src/intro.txt": MSG[0],
+            "src/outro.txt": MSG[1],
+            "src/script": (
+                "$mkdir res && $cat *.txt > message.txt && $cat message.txt | $tee res/log.txt"
+            ),
+        }
+    )
+
+    RES = "".join(MSG)
+    assert_gen_rule_result(
+        rule_runner,
+        Address("src", target_name="hello"),
+        expected_contents={
+            "src/message.txt": RES,
+            "src/res/log.txt": RES,
+        },
+    )
+
+
+def test_quotes_command(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "src/BUILD": dedent(
+                """\
+                gen_rule(
+                  name="quotes",
+                  tools=["echo", "tee"],
+                  command='$echo "foo bar" | $tee out.log',
+                  outputs=["out.log"],
+                )
+                """
+            ),
+        }
+    )
+
+    assert_gen_rule_result(
+        rule_runner,
+        Address("src", target_name="quotes"),
+        expected_contents={"src/out.log": "foo bar\n"},
+    )
+
+
+def test_chained_gen_rules(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "src/a/BUILD": dedent(
+                """\
+                gen_rule(
+                  name="msg",
+                  tools=["echo"],
+                  outputs=["msg"],
+                  command="$echo 'gen_rule:a' > msg",
+                )
+                """
+            ),
+            "src/b/BUILD": dedent(
+                """\
+                gen_rule(
+                  name="msg",
+                  tools=["cp", "echo"],
+                  outputs=["msg"],
+                  command="$cp ../a/msg . ; $echo 'gen_rule:b' >> msg",
```

Comment:
I had a bit of struggle to get this right, as I was confused about where msg a's input file landed.

Reply:
Yea, I think that finding the right semantics for this will definitely take some iteration. It looks like you've really done a great job experimenting here (thank you!), but probably safest to give the target an `experimental_` name until we've gotten more feedback.